horovod/common/operations.cc

```diff
@@ -1487,11 +1339,251 @@ void BackgroundThreadLoop(HorovodGlobalState& state) {
+//
+//      c) The coordinator receives the MPIRequests from the workers, as well
+//      as from its own TensorFlow ops, and stores them in a request table. The
+//      coordinator continues to receive MPIRequest messages until it has
+//      received MPI_SIZE number of empty "DONE" messages.
+//
+//      d) The coordinator finds all tensors that are ready to be reduced,
+//      gathered, or all operations that result in an error. For each of those,
+//      it sends an MPIResponse to all the workers. When no more MPIResponses
+//      are available, it sends a "DONE" response to the workers. If the process
+//      is being shutdown, it instead sends a "SHUTDOWN" response.
+//
+//      e) The workers listen for MPIResponse messages, processing each one by
+//      doing the required reduce or gather, until they receive a "DONE"
+//      response from the coordinator. At that point, the tick ends.
+//      If instead of "DONE" they receive "SHUTDOWN", they exit their background
+//      loop.
+bool RunLoopOnce(HorovodGlobalState& state, bool is_coordinator) {
+  // The coordinator sends a SHUTDOWN message to trigger shutdown.
+  bool should_shut_down = false;
+
+  // This delay determines thread frequency and MPI message latency
+  auto sleep_duration =
+      state.last_cycle_start +
+      std::chrono::microseconds(long(state.cycle_time * 1000.)) -
+      std::chrono::steady_clock::now();
+  if (sleep_duration > std::chrono::steady_clock::duration::zero()) {
+    std::this_thread::sleep_for(sleep_duration);
+  }
+  state.last_cycle_start = std::chrono::steady_clock::now();
+
+  // Copy the data structures from global state under this lock.
+  // However, don't keep the lock for the rest of the loop, so that
+  // enqueued stream callbacks can continue.
+  std::queue<MPIRequest> message_queue;
+  {
+    std::lock_guard<std::mutex> guard(state.mutex);
+    while (!state.message_queue.empty()) {
+      MPIRequest message = state.message_queue.front();
+      state.message_queue.pop();
+      message_queue.push(message);
+    }
+  }
+
+  // Collect all tensors that are ready to be reduced. Record them in the
+  // tensor count table (rank zero) or send them to rank zero to be
+  // recorded (everyone else).
+  std::vector<std::string> ready_to_reduce;
+  if (is_coordinator) {
+    while (!message_queue.empty()) {
+      // Pop the first available message message
+      MPIRequest message = message_queue.front();
+      message_queue.pop();
+
+      bool reduce =
+          IncrementTensorCount(state.message_table, message, state.size);
+      if (reduce) {
+        ready_to_reduce.push_back(message.tensor_name());
+      }
+    }
+
+    // Rank zero has put all its own tensors in the tensor count table.
+    // Now, it should count all the tensors that are coming from other
+    // ranks at this tick.
+
+    // 1. Get message lengths from every rank.
+    auto recvcounts = new int[state.size];
```

Comment:
There seem to be a lot of mallocs in this function.  How often is this routine executed?  (I see it sleeps some duration between ticks).  Might it be worth considering keeping some of these lists as globals that dynamically grow (not sure which resource is more constrained here, memory or compute)?

Reply:
These tiny mallocs are most likely not an issue, but [this malloc](https://github.com/uber/horovod/pull/329/files#diff-98391bc0c5b3bf87519e745b4ec3b9fdR923) could be quite expensive.  Determining tradeoff between such vanilla malloc, and cudaHostAlloc with reusable buffer is in the scope of future work.